batch-processing/pipeline_test.go

```diff
@@ -0,0 +1,795 @@
+			wg.Wait()
+			close(convertOut)
+
+			var gotArr []*ConvertOut
+
+			for v := range convertOut {
+				gotArr = append(gotArr, v)
+			}
+
+			require.Equal(t, len(gotArr), len(test.In))
+
+			require.ElementsMatch(t, test.ExpectedCalls, got)
+
+			require.ElementsMatch(t, test.ExpectedOut, gotArr)
+		})
+	}
+
+	t.Run("erroneous converter test", func(t *testing.T) {
+		p := NewPipeline(&Config{
+			Converter: &Converter{
+				cmd: func(a []string) (*imagick.ImageCommandResult, error) {
+					return nil, fmt.Errorf("some error")
+				},
+			},
+		})
+
+		p.errOut = make(chan *ErrOut, 1)
+
+		downOut := make(chan *DownloadOut, 1)
+		convertOut := make(chan *ConvertOut, 1)
+
+		downOut <- &DownloadOut{
+			Url: "http://localhost:8080/test1.jpg",
+			Key: "test1",
+			Ext: "jpg",
+		}
+
+		close(downOut)
+
+		wg := &sync.WaitGroup{}
+		wg.Add(1)
+		go p.Convert(wg, downOut, convertOut)
+		wg.Wait()
+		close(convertOut)
+
+		var gotArr []*ConvertOut
+
+		for v := range convertOut {
+			gotArr = append(gotArr, v)
+		}
+
+		expErr := &ErrOut{
+			Url: "http://localhost:8080/test1.jpg",
+			Key: "test1",
+			Ext: "jpg",
+			Err: fmt.Errorf("error converting image: some error"),
+		}
+
+		require.Equal(t, 0, len(gotArr), "should not have any output")
+
+		require.Equal(t, expErr, <-p.errOut)
+	})
+}
+
+func TestPipeline_Upload(t *testing.T) {
+	call := &s3.PutObjectInput{}
+	mockPutObject := func(input *s3.PutObjectInput) (*s3.PutObjectOutput, error) {
+		call = input
+		return &s3.PutObjectOutput{}, nil
+	}
+
+	p := NewPipeline(&Config{
+		Aws: &AWSConfig{
+			s3bucket:  "some_bucket",
+			PutObject: mockPutObject,
+		},
+	})
+
+	p.maxRetries = 1
+	p.tmpWorkingDir = "/inputs/test_assets"
+
+	tests := map[string]struct {
+		In             []*ConvertOut
+		ExpectedCall   *s3.PutObjectInput
+		ExpectedOut    []*UploadOut
+		ExpectedErrOut []*ErrOut
+	}{
+		"valid file": {
+			In: []*ConvertOut{
+				{
+					Url: "http://someurl:8080/test.jpg",
+					Key: "test",
+					Ext: "jpg",
+				},
+			},
+			ExpectedCall: &s3.PutObjectInput{
+				Bucket: aws.String("some_bucket"),
+				Key:    aws.String("test-converted.jpg"),
+			},
+			ExpectedOut: []*UploadOut{
+				{
+					Url:   "http://someurl:8080/test.jpg",
+					Key:   "test",
+					Ext:   "jpg",
+					S3url: "https://some_bucket.s3.amazonaws.com/test-converted.jpg",
+				},
+			},
+		},
+		"erroneous file": {
+			In: []*ConvertOut{
+				{
+					Url: "http://someurl:8080/not-found.jpg",
+					Key: "not-found",
+					Ext: "jpg",
+				},
+			},
+			ExpectedCall: &s3.PutObjectInput{},
+			ExpectedErrOut: []*ErrOut{
+				{
+					Url: "http://someurl:8080/not-found.jpg",
+					Key: "not-found",
+					Ext: "jpg",
+					Err: fmt.Errorf("some error"),
+				},
+			},
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			convertOut := make(chan *ConvertOut, len(test.In))
+			uploadOut := make(chan *UploadOut, len(test.In))
+
+			defer t.Cleanup(func() {
```

Comment:
You don't yourself need to `defer` this - `Cleanup` automatically registers the function to be effectively deferred itself.